src/care/array_utils.h

```diff
@@ -610,6 +633,26 @@ CARE_HOST_DEVICE inline int BinarySearch(const T *map, const int start,
    }
 }
 
+#if !defined(CARE_ENABLE_IMPLICIT_CONVERSIONS)
+template<typename mapType>
+CARE_HOST_DEVICE inline int BinarySearch(const care::host_device_ptr<const mapType>& map, const int start,
+                                         const int mapSize, const mapType num,
+                                         bool returnUpperBound)
+{
+#if defined(__CUDA_ARCH__) || defined(__HIP_DEVICE_COMPILE__) 
+   return BinarySearch<const mapType>(map.data(), start, mapSize, num, returnUpperBound);
+#else
+   int result = -1;
+   LOOP_SEQUENTIAL_REF(i, 0, 1, result) {
```

Comment:
I don't see why you need separate behavior for host vs. device. This function should just be a wrapper, so `return BinarySearch<const mapType>(map.data(), start, mapSize, num, returnUpperBound);` should be sufficient.

Reply:
Fixed